### PR TITLE
feat: Add .waspignore file

### LINKED ISSUE #192: Introduce "ignore" file for files that shouldn't be copied from ext/
There is no need to try to copy e.g. tmp editor files (at the end they don't get copied because they usually disappear), so we could introduce something like our "gitignore" for the usual set of files to be ignored (vim, emacs, intellij, ...). User could also expand it with their own file extension.

### PR DESCRIPTION
**Pull request by craigmc08**
# .waspignore File

Adds a new feature: the `.waspignore` file. The file exists in the root directory of a wasp project and contains a list of glob patterns, comment lines, and blank lines. Any files matching these glob patterns in the `ext/` directory will be ignored. A default `.waspignore` is created by the `wasp new` command that ignores some editor tmp files.

This PR introduces a new dependency, the [Glob](https://hackage.haskell.org/package/Glob) package. This package is, unfortunately, unmaintained currently but is the best option for glob matching I could find.

The default `.waspignore` file was also added to the example projects.

Fixes #192

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

### CODE REVIEW

File: waspc/test/WaspIgnoreFileTest.hs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+module WaspIgnoreFileTest where
```
Comment (Martinsos) [inline waspc/test/WaspIgnoreFileTest.hs:1]: Cool, I love the tests, you even used QuickCheck, nice :)!!

**Review by Martinsos — COMMENTED** [review]
Great job, this is really well done! Comments, tests, structure -> nicely done.
I left some minor comments so please take a look.

There is only one bigger thing I think we should figure out before we call this done, and that is how do we explain in the codebase (and to users) that .waspignore is ignoring files from ext/.
First of all, we ourselves have to agree on it: .waspignore is ignoring files from ext/. Ok, that makes sense, because that is what we need right now. What I find confusing is that .waspignore is in the root of the project, and not in the ext/ dir. Because looking at it, I would never guess the paths are relative to ext/. Hm! So, I think that is the problem. I see two solutions:
1. .waspignore has to be in ext/ dir.
2. .waspignore stays in root of the project, but then paths also refer to root of the project. But that really complicates things hm, question is then which other files do we ignore (we don't), and people need to put ext/ in front,. and we need to handle it differently in the code and so on.

Therefore, I would just put it in the ext/ dir (did I say at the beginning of this PR that we should put it in root and not in ext/? If so sorry, I realize now this is a better option).

Finally, I think the question remaining in the codebase is do we still call it WaspignoreFile, if it is specialized for ext/ dir files? Should it instead be called ExternalCode.WaspignoreFile? Hmmm. I would probably let it be as it is, just WaspignoreFile, but I would remove it from CompileOptions and just calculate its path as <pathToExtDir>/.waspignore .

**Comment by craigmc08** [comment]
@Martinsos Renaming everything to use a lowercase 'i' makes sense, and so does keeping the comment style consistent.

Moving `.waspignore` to `ext/` does seem to make more intuitive sense, it will be more clear what is being ignored. `.waspignore` itself should probably be ignored. Does it make more sense to make this an implicit rule (implemented in source and described in docs) or should the default `.waspignore` file list itself as a pattern? I'm personally leaning towards the implicit rule, but I don't have a good reason.

I'm happy with removing `waspignoreFilePath` from `CompileOptions`, it felt like unnecessary plumbing to get around not knowing where the wasp root dir was in `ExternalCode.readFiles`. Maybe the code was hinting to move the ignore file to `ext/` :)

**Comment by Martinsos** [comment]
> @Martinsos Renaming everything to use a lowercase 'i' makes sense, and so does keeping the comment style consistent.
> 
> Moving `.waspignore` to `ext/` does seem to make more intuitive sense, it will be more clear what is being ignored. `.waspignore` itself should probably be ignored. Does it make more sense to make this an implicit rule (implemented in source and described in docs) or should the default `.waspignore` file list itself as a pattern? I'm personally leaning towards the implicit rule, but I don't have a good reason.
> 
> I'm happy with removing `waspignoreFilePath` from `CompileOptions`, it felt like unnecessary plumbing to get around not knowing where the wasp root dir was in `ExternalCode.readFiles`. Maybe the code was hinting to move the ignore file to `ext/` :)

Good point, it probably was hinting that :).

Great, all sounds good! I agree with the implicit approach, and I think there is a good reason -> there is just no situation where user might want to not-ignore .waspignore file, so why complicate it for them by giving them flexibility they don't need and just allowing them to shoot themselves into the leg? So I think your intuition here is right.

Yay, let's do it!

**Comment by craigmc08** [comment]
Finished those changes, ready for them to be reviewed

**Comment by Martinsos** [comment]
Awesome @craigmc08 , this is it! Merging :).
Next step would be to add some documentation about this -> would you be up for that?
That should go in other repo, wasp-lang/web. Where exactly -> probably under https://wasp-lang.dev/docs/language/basic-elements, where you could create "External Code" header and under it you could create ".waspignore" header shortly describing how .waspignore works.

**Pull request merged**